src/strategy/rejection.rs

```diff
@@ -0,0 +1,196 @@
+//-
+// Copyright 2017 The proptest developers
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Provides `Rejection`, a compact representation of why something, such
+//! as a generated value, was rejected.
+
+use std::borrow::Borrow;
+use std::cmp::Ordering;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+use std::rc::Rc;
+
+use self::Rejection::*;
+
+/// The reason for why something, such as a generated value, was rejected.
+///
+/// # Representation
+/// 
+/// This type is currently representationally equivalent to:
+///
+/// ```rust
+/// use std::rc::Rc;
+/// 
+/// pub enum Rejection {
+///     Borrowed(&'static str),
+///     BoxOwned(Box<str>),
+///     RcOwned(Rc<str>),
+/// }
+///
+/// # fn main() {
+/// # use std::mem::size_of;
+/// assert_eq!(size_of::<Rejection>(), 3 * size_of::<usize>());
+/// # }
+/// ```
+///
+/// This is a compact and essentially copy-on-write representation while also
+/// allowing you not to allocate if the rejection message is not dynamic in
+/// nature. This is most efficient and should be used when possible. If you
+/// are allocating and then cloning the allocated `Rejection`, then you should
+/// prefer the `Rejection::RcOwned` variant as it will only bump the reference
+/// count and not actually do any new heap allocation.
+pub enum Rejection {
```

Comment:
This is eerily similar in functionality to [Supercow](https://docs.rs/supercow/0.1.0/supercow/). I'm not sure whether we'd want to bring that into proptest, just noting.

Reply:
The `Rc` part? What would be the downside? Carrying around `String` is also 3 words but we don't use the `capacity` word for anything atm. Removing `Rc` still means that `Rejection` is 3 words, but you loose the ability to cheaply clone. I think `Rc` is primarily useful in `Filter` where cloning of the string can happen frequently.

Reply:
> The Rc part?

The whole thing really. Supercow is a reference + box + arc/rc/etc smart pointer, `Rejection` is reference + box + rc smart pointer.